### DISCUSSION TITLE
Please Help: GitHub Action Test Code Build Failure

### DISCUSSION THREAD

**Discussion by ohsutinn** [community/community#125525] [Programming Help]
### Body

```
@TestPropertySource(locations = "/application-test.properties")
@SpringBootTest
class CouponConcurrencyTest {

    @Autowired
    private CouponService couponService;

    @Autowired
    private RedisLockStockFacade redisLockStockFacade;

    @Autowired
    private UserCouponRepository userCouponRepository;

    @Autowired
    private UserRepository userRepository;

    CouponIssueParam param;

    List<User> users;

    @BeforeEach
    void setUp() {
        param = new CouponIssueParam(1L);

        users = userRepository.findAll();
    }

    @AfterEach
    void after() {
        userCouponRepository.deleteAll();
    }

    @Test
    @Transactional
    @DisplayName("one_Issue_coupon")
    void one_Issue_coupon() {
        // given
        User user = userRepository.findById(1L).get();

        // when
        couponService.issueCoupon(param, user);

        // then
        int count = userCouponRepository.countByCouponId(param.getCouponId());
        AssertionsForClassTypes.assertThat(count).isEqualTo(1L);
    }
}
```

The test code mentioned above typically references an `application-test.properties` file in the test class setup.

```
spring.application.name=Issue-Coupon-Service

spring.datasource.url=jdbc:mysql://localhost:3306/test_coupon
spring.datasource.username=root
spring.datasource.password=1234
spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver

spring.jpa.properties.hibernate.dialect=org.hibernate.dialect.MySQL8Dialect

spring.jpa.hibernate.ddl-auto=update

spring.jpa.properties.hibernate.show_sql=true
spring.jpa.properties.hibernate.format_sql=true
```

`application-test.properties` file is written as described above.

```
name: Java CI with Gradle

on:
  push:
    branches: [ "main" ]
  pull_request:
    branches: [ "main" ]

permissions: write-all

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v3
      - name: Set up JDK 17
        uses: actions/setup-java@v3
        with:
          java-version: '17'
          distribution: 'adopt'
      
      - name: Setup MySQL
        uses: samin/mysql-action@v1
        with:
          mysql version: '8.0.33' 
          mysql root password: '1234' 
          mysql database: 'test_coupon'  
          mysql user: 'root'  
          mysql password: 'dbsehdwn12!' 
          character set server: 'utf8'  
          collation server: 'utf8_unicode_ci'  

      # Execut Gradle
      - name: Run chmod to make gradlew executable
        run: chmod +x ./gradlew

      # Bulid Test
      - name: Build with Gradle
        uses: gradle/gradle-build-action@bd5760595778326ba7f1441bcf7e88b49de61a25 # v2.6.0
        with:
          arguments: build


      - name: Publish Test Results
        uses: EnricoMi/publish-unit-test-result-action@v2
        if: always()
        with:
          files: |
            **/build/test-results/**/*.xml


      - name: Publish Test Report
        uses: mikepenz/action-junit-report@v3
        if: success() || failure() 
        with:
          report_paths: '**/build/test-results/test/TEST-*.xml'
```

`GitHub Action` is configured as mentioned above.

It seems like everything that needed to be set up has been done.

```
java.lang.IllegalStateException: Failed to load ApplicationContext for [WebMergedContextConfiguration@7048d039 testClass = com.coupon.issuecouponservice.service.coupon.CouponConcurrencyTest, locations = [], classes = [com.coupon.issuecouponservice.IssueCouponServiceApplication], contextInitializerClasses = [], activeProfiles = [], propertySourceDescriptors = [PropertySourceDescriptor[locations=[classpath:/application-test.properties], ignoreResourceNotFound=false, name=null, propertySourceFactory=null, encoding=null]], propertySourceProperties = ["org.springframework.boot.test.context.SpringBootTestContextBootstrapper=true"], contextCustomizers = [org.springframework.boot.test.autoconfigure.actuate.observability.ObservabilityContextCustomizerFactory$DisableObservabilityContextCustomizer@1f, org.springframework.boot.test.autoconfigure.properties.PropertyMappingContextCustomizer@0, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizer@4ce7fffa, org.springframework.boot.test.context.filter.ExcludeFilterContextCustomizer@6bbe2511, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContextCustomizer@4dd02341, org.springframework.boot.test.mock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.web.client.TestRestTemplateContextCustomizer@6399551e, org.springframework.boot.test.context.SpringBootTestAnnotation@42bb9565], resourceBasePath = "src/main/webapp", contextLoader = org.springframework.boot.test.context.SpringBootContextLoader, parent = null]
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:180)
 ```

```
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'entityManagerFactory' defined in class path resource [org/springframework/boot/autoconfigure/orm/jpa/HibernateJpaConfiguration.class]: [PersistenceUnit: default] Unable to build Hibernate SessionFactory; nested exception is org.hibernate.exception.JDBCConnectionException: Unable to open JDBC Connection for DDL execution [Communications link failure
 ```

```
Caused by: jakarta.persistence.PersistenceException: [PersistenceUnit: default] Unable to build Hibernate SessionFactory; nested exception is org.hibernate.exception.JDBCConnectionException: Unable to open JDBC Connection for DDL execution [Communications link failure
```

```
The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server.] [n/a]
 ```

```
Caused by: org.hibernate.exception.JDBCConnectionException: Unable to open JDBC Connection for DDL execution [Communications link failure
 ```

```
Caused by: com.mysql.cj.jdbc.exceptions.CommunicationsException: Communications link failure
 ```

```
Caused by: com.mysql.cj.exceptions.CJCommunicationsException: Communications link failure
 ```

```
Caused by: java.net.ConnectException: Connection refused
```

I keep getting the errors mentioned above. 😭😭😭😭😭😭

What could possibly be the reason?

I haven't set up a separate test server; the tests run fine just in the local environment.

I can't figure out the cause...

Please translate this as well.

### Guidelines

- [ ] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by mranv** [comment]
The error message you're receiving, `java.net.ConnectException: Connection refused`, indicates that your application is unable to connect to the MySQL database. This issue typically occurs when the MySQL server is not running or is not accessible from the application.

Here are some possible reasons for this error and potential solutions:

1. **MySQL Server Not Running**: Ensure that the MySQL server is running on the machine where the tests are being executed. If you're running the tests locally, make sure the MySQL server is up and running. If you're running the tests on a CI/CD server or a different machine, you'll need to ensure that the MySQL server is installed and running on that machine.

2. **Incorrect MySQL Connection Details**: Double-check the MySQL connection details in your `application-test.properties` file. Ensure that the `spring.datasource.url`, `spring.datasource.username`, and `spring.datasource.password` are correct and match the MySQL server's configuration.

3. **Firewall/Network Configuration**: If the MySQL server is running on a different machine or a remote server, you may need to check the firewall settings and network configuration to ensure that the application can connect to the MySQL server over the network.

4. **MySQL Server Accessibility**: Verify that the MySQL server is accessible from the machine where the tests are being executed. You can try connecting to the MySQL server using a MySQL client or a database administration tool from the same machine to ensure that the server is reachable.

5. **Permissions and Privileges**: Check if the MySQL user account you're using has the necessary permissions and privileges to connect to the database and perform the required operations.

To troubleshoot further, you can try the following steps:

1. Verify that the MySQL server is running by checking its status or trying to connect to it from the same machine using a MySQL client or a database administration tool.

2. Update the `application-test.properties` file with the correct MySQL connection details and ensure that the credentials have the necessary permissions.

3. If the MySQL server is running on a different machine or a remote server, check the network configuration and firewall settings to ensure that the application can connect to the MySQL server over the network.

4. If you're running the tests on a CI/CD server or a different machine, ensure that the MySQL server is installed and running on that machine, and update the connection details accordingly.

5. Try connecting to the MySQL server from the machine where the tests are being executed using a MySQL client or a database administration tool to verify that the connection can be established.

If the issue persists after trying these steps, you may need to provide more details about your setup, such as the environment where the tests are being executed (local machine, CI/CD server, etc.), the MySQL server configuration, and any additional network or security configurations that might be relevant.
